File: agenttel-node/src/genai/span-builder.ts
import { trace, type Span, type Tracer, SpanKind, context as otelContext } from '@opentelemetry/api';
import * as attrs from '../attributes.js';
import { calculateCost } from './cost.js';

export interface ChatResult {
  responseModel: string;
  responseId: string;
  inputTokens: number;
  outputTokens: number;
  finishReasons?: string[];
}

export interface TokenDetails {
  reasoningTokens?: number;
  cachedReadTokens?: number;
  cachedWriteTokens?: number;
}

/** Creates properly-formed GenAI spans following OTel semantic conventions. */
export class GenAiSpanBuilder {
  private readonly tracer: Tracer;

  constructor(tracerOrName?: Tracer | string) {
    this.tracer = typeof tracerOrName === 'string'
      ? trace.getTracer(tracerOrName)
      : tracerOrName ?? trace.getTracer('agenttel.genai');
  }

  /** Create a new GenAI chat span. */
  startChatSpan(model: string, system: string): Span {
    const span = this.tracer.startSpan(`chat ${model}`, {
      kind: SpanKind.CLIENT,
      attributes: {
        [attrs.GENAI_OPERATION_NAME]: 'chat',
        [attrs.GENAI_SYSTEM]: system,
        [attrs.GENAI_REQUEST_MODEL]: model,
      },
    });
    return span;
  }

  /** Set optional request parameters on a span. */
  static setRequestParams(span: Span, temperature?: number, maxTokens?: number, topP?: number): void {
    if (temperature !== undefined) span.setAttribute(attrs.GENAI_REQUEST_TEMPERATURE, temperature);
    if (maxTokens !== undefined) span.setAttribute(attrs.GENAI_REQUEST_MAX_TOKENS, maxTokens);
    if (topP !== undefined) span.setAttribute(attrs.GENAI_REQUEST_TOP_P, topP);
  }

  /** Complete a GenAI span with response data and cost calculation. */
  static endSpanSuccess(span: Span, result: ChatResult): void {
    span.setAttribute(attrs.GENAI_RESPONSE_MODEL, result.responseModel);
    span.setAttribute(attrs.GENAI_RESPONSE_ID, result.responseId);
    span.setAttribute(attrs.GENAI_USAGE_INPUT_TOKENS, result.inputTokens);
    span.setAttribute(attrs.GENAI_USAGE_OUTPUT_TOKENS, result.outputTokens);

    if (result.finishReasons?.length) {
      span.setAttribute(attrs.GENAI_RESPONSE_FINISH_REASONS, result.finishReasons);
    }

    const cost = calculateCost(result.responseModel, result.inputTokens, result.outputTokens);
    if (cost > 0) {
      span.setAttribute(attrs.GENAI_COST_USD, cost);
    }

    span.end();
  }

  /** Complete a GenAI span with an error. */
  static endSpanError(span: Span, error: Error): void {
    span.recordException(error);
    span.end();
  }

  /** Set extended token usage details. */
  static setTokenDetails(span: Span, details: TokenDetails): void {
    if (details.reasoningTokens) span.setAttribute(attrs.GENAI_REASONING_TOKENS, details.reasoningTokens);
    if (details.cachedReadTokens) span.setAttribute(attrs.GENAI_CACHED_READ_TOKENS, details.cachedReadTokens);
    if (details.cachedWriteTokens) span.setAttribute(attrs.GENAI_CACHED_WRITE_TOKENS, details.cachedWriteTokens);
  }

  /** Set time-to-first-token metric. */
  static setTimeToFirstToken(span: Span, ttftMs: number): void {
    span.setAttribute(attrs.GENAI_TIME_TO_FIRST_TOKEN_MS, ttftMs);
  }

  /** Set the GenAI framework attribute. */
  static setFramework(span: Span, framework: string): void {
    span.setAttribute(attrs.GENAI_FRAMEWORK, framework);
  }
}

File: agenttel-go/genai/span_builder.go
// Package genai provides GenAI span builders and cost calculators for LLM instrumentation.
package genai

import (
	"context"
	"fmt"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/trace"

	attrs "go.agenttel.dev/agenttel/attributes"
)

// SpanBuilder creates properly-formed GenAI spans following OTel semantic conventions.
type SpanBuilder struct {
	tracer trace.Tracer
}

// NewSpanBuilder creates a GenAI span builder.
func NewSpanBuilder(tracer trace.Tracer) *SpanBuilder {
	return &SpanBuilder{tracer: tracer}
}

// ChatResult holds the response data for ending a chat span.
type ChatResult struct {
	ResponseModel string
	ResponseID    string
	InputTokens   int64
	OutputTokens  int64
	FinishReasons []string
}

// TokenDetails holds extended token usage data.
type TokenDetails struct {
	ReasoningTokens  int64
	CachedReadTokens int64
	CachedWriteTokens int64
}

// StartChatSpan creates a new GenAI chat span.
func (b *SpanBuilder) StartChatSpan(ctx context.Context, model, system string) (context.Context, trace.Span) {
	spanName := fmt.Sprintf("chat %s", model)
	ctx, span := b.tracer.Start(ctx, spanName, trace.WithSpanKind(trace.SpanKindClient))

	span.SetAttributes(
		attribute.String(attrs.GenAIOperationName, "chat"),
		attribute.String(attrs.GenAISystem, system),
		attribute.String(attrs.GenAIRequestModel, model),
	)

	return ctx, span
}

// SetRequestParams sets optional request parameters on a span.
func SetRequestParams(span trace.Span, temperature float64, maxTokens int64, topP float64) {
	if temperature > 0 {
		span.SetAttributes(attribute.Float64(attrs.GenAIRequestTemperature, temperature))
	}
	if maxTokens > 0 {
		span.SetAttributes(attribute.Int64(attrs.GenAIRequestMaxTokens, maxTokens))
	}
	if topP > 0 {
		span.SetAttributes(attribute.Float64(attrs.GenAIRequestTopP, topP))
	}
}

// EndSpanSuccess completes a GenAI span with response data and cost calculation.
func EndSpanSuccess(span trace.Span, result ChatResult) {
	span.SetAttributes(
		attribute.String(attrs.GenAIResponseModel, result.ResponseModel),
		attribute.String(attrs.GenAIResponseID, result.ResponseID),
		attribute.Int64(attrs.GenAIUsageInputTokens, result.InputTokens),
		attribute.Int64(attrs.GenAIUsageOutputTokens, result.OutputTokens),
	)

	if len(result.FinishReasons) > 0 {
		span.SetAttributes(attribute.StringSlice(attrs.GenAIResponseFinishReasons, result.FinishReasons))
	}

	// Calculate cost
	cost := CalculateCost(result.ResponseModel, result.InputTokens, result.OutputTokens)
	if cost > 0 {
		span.SetAttributes(attribute.Float64(attrs.GenAICostUSD, cost))
	}

	span.End()
}

// EndSpanError completes a GenAI span with an error.
func EndSpanError(span trace.Span, err error) {
	span.RecordError(err)
	span.End()
}

// SetTokenDetails sets extended token usage details.
func SetTokenDetails(span trace.Span, details TokenDetails) {
	if details.ReasoningTokens > 0 {
		span.SetAttributes(attribute.Int64(attrs.GenAIReasoningTokens, details.ReasoningTokens))
	}
	if details.CachedReadTokens > 0 {
		span.SetAttributes(attribute.Int64(attrs.GenAICachedReadTokens, details.CachedReadTokens))
	}
	if details.CachedWriteTokens > 0 {
		span.SetAttributes(attribute.Int64(attrs.GenAICachedWriteTokens, details.CachedWriteTokens))
	}
}

// SetTimeToFirstToken sets the time-to-first-token metric.
func SetTimeToFirstToken(span trace.Span, ttftMs int64) {
	span.SetAttributes(attribute.Int64(attrs.GenAITimeToFirstTokenMs, ttftMs))
}

// SetFramework sets the GenAI framework attribute.
func SetFramework(span trace.Span, framework string) {
	span.SetAttributes(attribute.String(attrs.GenAIFramework, framework))
}
